c/tskit/trees.c

```diff
@@ -7814,75 +7835,455 @@ remove_unextended(edge_list_t **head, edge_list_t **tail)
+    double *sites_position = tables->sites.position;
+    double *nodes_time = tables->nodes.time;
+    tsk_tree_t tree;
 
-    if (keep == NULL || out_parent == NULL || degree == NULL || not_sample == NULL) {
-        ret = TSK_ERR_NO_MEMORY;
+    ret = tsk_tree_init(&tree, self, TSK_NO_SAMPLE_COUNTS);
+    if (ret != 0) {
         goto out;
     }
-    tsk_memset(out_parent, 0xff, num_nodes * sizeof(*out_parent));
 
-    ret = tsk_blkalloc_init(&edge_list_heap, 8192);
+    next_mut = 0;
+    for (ret = tsk_tree_first(&tree); ret == TSK_TREE_OK; ret = tsk_tree_next(&tree)) {
+        while (next_mut < (tsk_id_t) mutations->num_rows
+               && sites_position[mutations->site[next_mut]] < tree.interval.right) {
+            t = mutations->time[next_mut];
+            if (tsk_is_unknown_time(t)) {
+                ret = TSK_ERR_DISALLOWED_UNKNOWN_MUTATION_TIME;
+                goto out;
+            }
+            c = mutations->node[next_mut];
+            tsk_bug_assert(c < (tsk_id_t) num_nodes);
+            p = tree.parent[c];
+            while (p != TSK_NULL && nodes_time[p] <= t) {
+                c = p;
+                p = tree.parent[c];
+            }
+            tsk_bug_assert(nodes_time[c] <= t);
+            mutations->node[next_mut] = c;
+            next_mut++;
+        }
+    }
     if (ret != 0) {
         goto out;
     }
-    ret = tsk_tree_position_init(&tree_pos, self, 0);
+
+out:
+    tsk_tree_free(&tree);
+
+    return ret;
+}
+
+typedef struct {
+    const tsk_treeseq_t *ts;
+    tsk_edge_table_t *edges;
+    int direction;
+    tsk_id_t *last_degree, *next_degree;
+    tsk_id_t *last_nodes_edge, *next_nodes_edge;
+    tsk_id_t *parent_out, *parent_in;
+    bool *not_sample;
+    double *near_side, *far_side;
+    edge_list_t *edges_out_head, *edges_out_tail;
+    edge_list_t *edges_in_head, *edges_in_tail;
+    tsk_blkalloc_t edge_list_heap;
+} haplotype_extender_t;
+
+static int
+haplotype_extender_init(haplotype_extender_t *self, const tsk_treeseq_t *ts,
+    int direction, tsk_edge_table_t *edges)
+{
+    int ret = 0;
+    tsk_id_t tj;
+    tsk_size_t num_nodes = tsk_treeseq_get_num_nodes(ts);
+
+    tsk_memset(self, 0, sizeof(haplotype_extender_t));
+
+    self->ts = ts;
+    self->edges = edges;
+    ret = tsk_edge_table_copy(&ts->tables->edges, self->edges, TSK_NO_INIT);
     if (ret != 0) {
         goto out;
     }
-    ret = tsk_edge_table_copy(&tables->edges, edges, TSK_NO_INIT);
+
+    self->direction = direction;
+    if (direction == TSK_DIR_FORWARD) {
+        self->near_side = self->edges->left;
+        self->far_side = self->edges->right;
+    } else {
+        self->near_side = self->edges->right;
+        self->far_side = self->edges->left;
+    }
+
+    self->edges_in_head = NULL;
+    self->edges_in_tail = NULL;
+    self->edges_out_head = NULL;
+    self->edges_out_tail = NULL;
+
+    tsk_memset(&self->edge_list_heap, 0, sizeof(self->edge_list_heap));
```

Comment:
You don't need to memset the edge_list_heap to 0 here as its already been done above when you zero'd the top level struct.